### PR TITLE
Madpack: Make install, reinstall and upgrade atomic

### PR DESCRIPTION
**Pull request by orhankislal**
We now write all the necessary sql into one file, and run it once in a
single session. The database's rollback will be useful to bring it back
to original state in case of a failure.

Co-authored-by: Rahul Iyer <riyer@apache.org>
Co-authored-by: Orhan Kislal <okislal@pivotal.io>

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/482/

**Comment by kaknikhil** [comment]
A few other observations from madpack output

1. While uninstalling , this message is printed on the console’
`madpack.py: INFO : Installing MADlib modules` We should instead say that we are uninstalling modules.

1. In case of failure, the output looks something like `MADlib upgrade unsuccessful.` It would be nice to also say that everything was rolled back.

1. I also noticed that the output messages have changed a bit for install/uninstall/reinstall and upgrade

old output 
```
madpack.py: INFO : Installing MADlib into MADLIB schema...
madpack.py: INFO : > Creating MADLIB schema
madpack.py: INFO : > Creating MADLIB.MigrationHistory table
madpack.py: INFO : > Writing version info in MigrationHistory table
madpack.py: INFO : > Creating objects for modules:
madpack.py: INFO : > - array_ops
madpack.py: INFO : > - bayes
...
...
madpack.py: INFO : > - validation
madpack.py: INFO : MADlib 1.13 installed successfully in MADLIB schema.
``` 

new output
```
madpack.py: INFO : Testing PL/Python environment...
madpack.py: INFO : > Creating language PL/Python...
madpack.py: INFO : > PL/Python environment OK (version: 2.7.14)
madpack.py: INFO : > Preparing objects for modules:
madpack.py: INFO : > - array_ops
madpack.py: INFO : > - bayes
...
...
madpack.py: INFO : > - validation
madpack.py: INFO : Installing MADlib modules...
madpack.py: INFO : > Created MADLIB schema
madpack.py: INFO : > Created MADLIB.MigrationHistory table
madpack.py: INFO : > Wrote version info in MigrationHistory table
madpack.py: INFO : MADlib 1.15-dev installed successfully in MADLIB schema.
```
I think the previous message was better because all the modules were printed after `Installing MADlib into MADLIB schema` instead of after `Preparing objects for modules`. And with the new output`Installing MADlib modules...` is followed by `Created madlib schema` which looks a bit weird.

**Comment by njayaram2** [comment]
Thank you for the comments @kaknikhil , will address them.

File: src/madpack/madpack.py [hunk 1]
```diff
@@ -824,6 +873,246 @@ def parse_arguments():
...
+            if madpack_cmd == 'reinstall':
+                info_(this, "Setting MADlib database version to be None for reinstall", verbose)
+                dbrev = None
+
+            info_(this, "*** Installing MADlib ***", True)
+
+            # 1) Compare OS and DB versions.
+            # noop if OS <= DB.
+            _print_revs(rev, dbrev, con_args, schema)
+            if is_rev_gte(get_rev_num(dbrev), get_rev_num(rev)):
```
[the remark this answers is not in the document]
Comment (njayaram2) [reply src/madpack/madpack.py]: Keeping the `if elif`, but changed the order. Although the checks are on different variables, they are directly related.
Renamed variables: `dbrev`->`db_madlib_ver`, and `rev`->`new_madlib_ver`.

File: src/madpack/madpack.py [hunk 2]
```diff
@@ -987,275 +1276,42 @@ def main(argv):
...
-            _internal_run_query("DROP SCHEMA IF EXISTS %s CASCADE;" % (test_schema), True)
-
-        # Drop install-check user
-        _internal_run_query("DROP OWNED BY %s CASCADE;" % (test_user), True)
-        _internal_run_query("DROP USER %s;" % (test_user), True)
+                if args.command[0] != 'uninstall':
+                    if args.command[0] == 'upgrade':
+                        info_(this, "MADlib %s upgraded successfully in %s schema." % (str(rev), schema.upper()), True)
+                    else:
+                        info_(this, "> Created %s schema" % schema.upper(), True)
```
[the remark this answers is not in the document]
Comment (njayaram2) [reply src/madpack/madpack.py]: Removed `upper()` function calls.

**Comment by njayaram2** [comment]
w/ @orhankislal  The reason we had to change the info messages that are printed during install/reinstall/upgrade is that we are now writing all the necessary SQL into a single file, and running it as an atomic operation. Previously, the messages were getting printed after individual modules were processed. If we keep the older info messages as is, it won't reflect what is happening underneath. Breaking the SQL file up to print messages will defeat the purpose of this PR.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/492/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/496/

**Pull request merged**